src/AutoRest.CSharp/Common/Output/Models/Types/ClientOptionsTypeProvider.cs

```diff
@@ -16,6 +17,7 @@ internal sealed class ClientOptionsTypeProvider : TypeProvider
 
         public FormattableString Description { get; }
         public IReadOnlyList<ApiVersion>? ApiVersions { get; }
+        public IList<Parameter>? AdditionalProperties { get; private set; }
```

Comment:
Is there any special meaning of this being null? An empty list should work perfectly.
Also we should not make this mutable - it is unsafe that the state of this provider could change. If it could change, how could we know it is the final state when we write it?

Reply:
I hope it could be immutable as well. However, when we built `LowLevelClient` we need to pass `clientOptions`, and the optional parameters needed by `clientOptions` comes from `LowLevelClient`. We create `clientOptions` way earlier than we create `LowLevelClient`.

Reply:
This is not a proper reason.
Something like this should work:
```
var additionalProperties = new List<Parameter>();
var options = new ClientOptionsTypeProvider(additionalProperties);

// build the client

// populate the array after we build the client
```
Or change this property to be lazy loaded so that when some one calls this, everything will be calculated and will never change.

Reply:
There is circle dependency. You cannot know additionalProperties before client is built.